test/users/npadmana/FFTW.chpl

```diff
@@ -28,39 +28,86 @@ module FFTW {
-      // Make sure types are correct
-      param ndim : c_int = dims.size;
-      return fftw_plan_dft(ndim, dims(1), c_ptrTo(in1) : _cxptr, c_ptrTo(out1) : _cxptr, sign, flags);
-    }
-
-  // Real-to-complex and complex-to-real plans
-  // We handle these with a type parameter to let the user pass in an appropriately sized
-  // real array for the complex part, most usually when doing an in-place transform.
-  proc plan_dft_r2c(dims, ref in1 : ?t, ref out1 : ?tt, flags : c_uint) : fftw_plan 
-    where ((t.type==real(64)) || (t.type==fftw_complex)) && ((tt.type==real(64)) || (tt.type==fftw_complex)) && (isHomogeneousTuple(dims))
-    {
-      //-- define the extern proc
-      extern proc fftw_plan_dft_r2c(rank : c_int, const ref n : c_int,  in1 : _rptr,  out1 : _cxptr, flags : c_uint) : fftw_plan;
-      // Make sure types are correct
-      param ndim : c_int = dims.size;
-      return fftw_plan_dft_r2c(ndim,dims(1),c_ptrTo(in1) : c_ptr(c_double), c_ptrTo(out1) : _cxptr, flags);
-    }
-  proc plan_dft_c2r(dims, ref in1 : ?t, ref out1 : ?tt, flags : c_uint) : fftw_plan 
-    where ((t.type==real(64)) || (t.type==fftw_complex)) && ((tt.type==real(64)) || (tt.type==fftw_complex)) && (isHomogeneousTuple(dims))
-    {
-      //-- define the extern proc
-      extern proc fftw_plan_dft_c2r(rank : c_int, const ref n : c_int, in1 : _cxptr,  out1 : _rptr, flags : c_uint) : fftw_plan;
-      // Make sure types are correct
-      param ndim : c_int = dims.size;
-      return fftw_plan_dft_c2r(ndim, dims(1),c_ptrTo(in1) : _cxptr, c_ptrTo(out1) : c_ptr(c_double), flags);
-    }
+  proc plan_dft(in1 : [] fftw_complex, out1 : [] fftw_complex, sign : c_int, flags :c_uint) : fftw_plan
+  {
+    param rank = in1.rank: c_int;
 
+    var dims: rank*c_int;
+    for param i in 1..rank do
+      dims(i) = in1.domain.dim(i).size: c_int;
 
+    extern proc fftw_plan_dft(rank: c_int, 
+        n,  // BLC: having trouble being specific
+        in1: [] fftw_complex, 
+        out1: [] fftw_complex, 
+        sign : c_int, flags : c_uint) : fftw_plan;
 
+    return fftw_plan_dft(rank, dims, in1, out1, sign, flags);
+  }
+
+  // Real-to-complex and complex-to-real planning routines
+  // There are two cases that we treat independently here : in-place and out-of-place transforms
+
+  // Since the calls to FFTW are the same, pull the extern declarations out.
+  // TODO : This should be cleaned up further and made consistent across the file
+  // TODO : NP : I'm dropping this back to pointers, because I don't know how to cast arrays.
+  extern proc fftw_plan_dft_r2c(rank: c_int, 
+      n,  // BLC: having trouble being specific
+      in1: _rptr,
+      out1: _cxptr, 
+      flags : c_uint) : fftw_plan;
+  extern proc fftw_plan_dft_c2r(rank: c_int, 
+      n,  // BLC: having trouble being specific
+      in1: _cxptr,
+      out1: _rptr,
+      flags : c_uint) : fftw_plan;
```

Comment:
I didn't have time to test either of these, but I'm thinking it should work to either make the extern routines take generic arrays:

  proc fftw_plan_dft_r2c(..., in1: [], out1: [], ...);

or to provide overloads for all the types of arrays we want to send as a fallback:

  proc fftw_plan_dft_r2c(..., in1: [] real, out1: [] complex, ...);
  proc fftw_plan_dft_r2c(..., in1: [] real, out1: [] real, ...);

Even though the C routines aren't generic/overloaded, the hypothesis is that this will tell the Chapel compiler what it needs to know to call the routines...

Or maybe my approach will still lead to C-compile-time issues?  Not certain.  I should have some time to look into this more tomorrow (meeting-free day).